### ISSUE TITLE
Fine-tuning with Naive Pipeline Parallel: NaN after optimizer step

### ISSUE THREAD

**Issue by fs4r** [zphang/minimal-llama#4]
Your model does not seem to be able to calculate the gradients of the layers correctly. When I run finetune_pp.py and print the loss during training, after the first optimizer step, the loss becomes the following: 

tensor(nan, device='cuda:1', dtype=torch.float16, grad_fn=<NllLossBackward0>)

Can you reproduce this on your machine? Otherwise, would you be willing to share your pip freeze, so that I can try out, if there is a package mismatch?

**Comment by fs4r** [comment]
This seems to be an overflow. I tried to solve it with torch autocast and grad scaler. The following seems to work:

Initializing the model without fp16 and deleting all .to(dtype) calls from the model definition.

```python
model = transformers.LLaMAForCausalLM.from_pretrained(
    args.model_path,
    torch_dtype=torch.float16,
)
```

And using autocast in the training routine

```python
with torch.autocast(device_type='cuda', dtype=torch.float16):
    logits = model_forward(input_ids, layer_device_tuples)
    loss = F.cross_entropy(
        logits.view(-1, model.config.vocab_size),
        labels.view(-1).to(logits.device),
    )
scaler.scale(loss).backward()
scaler.step(opt)
```

**Comment by yysjasmine** [comment]
I met the same situation, how did you fix it?

**Comment by chaoyi-wu** [comment]
I met the same situation
